packages/block-editor/src/hooks/color.js

```diff
@@ -0,0 +1,250 @@
+	if ( ! hasBlockSupport( settings, COLOR_SUPPORT_KEY ) ) {
+		return settings;
+	}
+	const existingGetEditWrapperProps = settings.getEditWrapperProps;
+	settings.getEditWrapperProps = ( attributes ) => {
+		let props = {};
+		if ( existingGetEditWrapperProps ) {
+			props = existingGetEditWrapperProps( attributes );
+		}
+		return addSaveProps( props, settings, attributes );
+	};
+
+	return settings;
+}
+
+const ColorPanel = ( { colorSettings, clientId } ) => {
+	const { getComputedStyle, Node } = window;
+
+	const [ detectedBackgroundColor, setDetectedBackgroundColor ] = useState();
+	const [ detectedColor, setDetectedColor ] = useState();
+
+	// TODO: check performance impact of running this on each render
+	useEffect( () => {
+		const colorsDetectionElement = getBlockDOMNode( clientId );
+		setDetectedColor( getComputedStyle( colorsDetectionElement ).color );
+
+		let backgroundColorNode = colorsDetectionElement;
+		let backgroundColor = getComputedStyle( backgroundColorNode )
+			.backgroundColor;
+		while (
+			backgroundColor === 'rgba(0, 0, 0, 0)' &&
+			backgroundColorNode.parentNode &&
+			backgroundColorNode.parentNode.nodeType === Node.ELEMENT_NODE
+		) {
+			backgroundColorNode = backgroundColorNode.parentNode;
+			backgroundColor = getComputedStyle( backgroundColorNode )
+				.backgroundColor;
+		}
+
+		setDetectedBackgroundColor( backgroundColor );
+	} );
+
+	return (
+		<InspectorControls>
+			<PanelColorSettings
+				title={ __( 'Color settings' ) }
+				initialOpen={ false }
+				colorSettings={ colorSettings }
+			>
+				<ContrastChecker
+					backgroundColor={ detectedBackgroundColor }
+					textColor={ detectedColor }
+				/>
+			</PanelColorSettings>
+		</InspectorControls>
+	);
+};
+
+/**
+ * Override the default edit UI to include new inspector controls for block
+ * color, if block defines support.
+ *
+ * @param  {Function} BlockEdit Original component
+ * @return {Function}           Wrapped component
+ */
+export const withBlockControls = createHigherOrderComponent(
+	( BlockEdit ) => ( props ) => {
+		const { name: blockName } = props;
+		const colors = useSelect( ( select ) => {
+			return select( 'core/block-editor' ).getSettings().colors;
+		}, [] );
+
+		if ( ! hasBlockSupport( blockName, COLOR_SUPPORT_KEY ) ) {
+			return <BlockEdit key="edit" { ...props } />;
+		}
+		const { style, textColor, backgroundColor } = props.attributes;
+
+		const onChangeColor = ( name ) => ( value ) => {
+			const colorObject = getColorObjectByColorValue( colors, value );
+			const attributeName = name + 'Color';
+			const newStyle = {
+				...style,
+				color: {
+					...style?.color,
+					[ name ]:
+						colorObject && colorObject.slug ? undefined : value,
+				},
+			};
+			const newNamedColor =
+				colorObject && colorObject.slug ? colorObject.slug : undefined;
+			props.setAttributes( {
+				style: newStyle,
+				[ attributeName ]: newNamedColor,
+			} );
+		};
+
+		return [
+			<ColorPanel
```

Comment:
There are cases that are not background color or text color, e.g: blocks may allow choosing border color and hoover colors:
![image](https://user-images.githubusercontent.com/11271197/77348094-c63d8600-6d30-11ea-9154-2e1743fcfcc5.png)


Other blocks have multiple text areas (multiple rich text elements) and have multiple color pickers for each text area plus color pickers for the separators:
![image](https://user-images.githubusercontent.com/11271197/77348975-3c8eb800-6d32-11ea-8bba-bf0a125151ad.png)


For these cases, the plan is to expand this API in the future? Or should they use the UI components plus useColors or the UI Components plus withColors?

Reply:
For me, probably a custom implementation using the reusable components. I'm still not sure about withColors or useColors, I'm not convinced by both. We'll see.

Reply:
I wouldn't worry about it too much at the moment. It will naturally evolve together with the development of other design-related features lead by @ItsJonQ. At some point, I expect that Global Styles and the block editor features API will intersect with this logic. This PR provides the best possible defaults that make the most sense as confirmed in the follow-up PRs. Once we introduce more color-related features like borders, hover and focus states, we could control them with block editor features per site or per block. 

> Other blocks have multiple text areas (multiple rich text elements) and have multiple color pickers for each text area plus color pickers for the separators:

It raises the question of whether they could be refactored to use nested blocks that would solve the issue.